src/ParallelAlgorithms/Interpolation/Callbacks/ObserveSurfaceData.hpp

```diff
@@ -0,0 +1,125 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/DataBox/TagName.hpp"
+#include "DataStructures/Tensor/TensorData.hpp"
+#include "IO/Observer/Tags.hpp"
+#include "IO/Observer/VolumeActions.hpp"
+#include "NumericalAlgorithms/SphericalHarmonics/Strahlkorper.hpp"
+#include "NumericalAlgorithms/SphericalHarmonics/Tags.hpp"
+#include "NumericalAlgorithms/SphericalHarmonics/YlmSpherepack.hpp"
+#include "Parallel/GlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Local.hpp"
+#include "Parallel/Reduction.hpp"
+#include "ParallelAlgorithms/Interpolation/InterpolationTargetDetail.hpp"
+#include "ParallelAlgorithms/Interpolation/Protocols/PostInterpolationCallback.hpp"
+#include "Utilities/Functional.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/PrettyType.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace db {
+template <typename TagsList>
+class DataBox;
+}  // namespace db
+namespace observers {
+template <class Metavariables>
+struct ObserverWriter;
+}  // namespace observers
+/// \endcond
+
+namespace intrp {
+namespace callbacks {
+
+/// \brief post_interpolation_callback that outputs
+/// 2D "volume" data on a surface.
+///
+/// Uses:
+/// - Metavariables
+///   - `temporal_id`
+/// - DataBox:
+///   - `TagsToObserve` (each tag must be a Scalar<DataVector>)
+///
+/// Conforms to the intrp::protocols::PostInterpolationCallback protocol
+///
+/// For requirements on InterpolationTargetTag, see
+/// intrp::protocols::InterpolationTargetTag
+template <typename TagsToObserve, typename InterpolationTargetTag>
+struct ObserveSurfaceData
+    : tt::ConformsTo<intrp::protocols::PostInterpolationCallback> {
+  static constexpr double fill_invalid_points_with =
+      std::numeric_limits<double>::quiet_NaN();
+
+  using const_global_cache_tags = tmpl::list<observers::Tags::SurfaceFileName>;
+
+  template <typename DbTags, typename Metavariables, typename TemporalId>
+  static void apply(const db::DataBox<DbTags>& box,
+                    Parallel::GlobalCache<Metavariables>& cache,
+                    const TemporalId& temporal_id) {
+    const Strahlkorper<Frame::Inertial>& strahlkorper =
+        get<StrahlkorperTags::Strahlkorper<Frame::Inertial>>(box);
+    const YlmSpherepack& ylm = strahlkorper.ylm_spherepack();
+    const std::array<DataVector, 2>& theta_phi = ylm.theta_phi_points();
+    const DataVector& theta = theta_phi[0];
```

Comment:
why not just inline the `theta_phi_points()` calls here? Then it's clear that `theta_phi` itself is never used

Reply:
I thought theta_phi_points() might be expensive, so I avoided calling it twice, but I'm happy to make this change

Reply:
Oh, I see. Could you make this not a `&` instead? That's what confused me. You're right, the result of `theta_phi_points()` is a value object not a reference, so there's some amount of computation done there

Reply:
@nilsdeppe  just to confirm: you're saying no & line 73, right? Or did you mean no & on line 74 or somewhere else?

Reply:
Yes, that's right, just line 73:
```cpp
     const std::array<DataVector, 2> theta_phi = ylm.theta_phi_points(); 
```